src/Internal/Service/DuplicatePaymentPreventionService.php

```diff
@@ -0,0 +1,260 @@
+ */
+class DuplicatePaymentPreventionService {
+	/**
+	 * Key name for saving the current processing order_id to WC Session with the purpose
+	 * of preventing duplicate payments in a single order.
+	 *
+	 * @type string
+	 */
+	const SESSION_KEY_PROCESSING_ORDER = 'wcpay_processing_order';
+
+	/**
+	 * Flag to indicate that a previous order with the same cart content has already paid.
+	 *
+	 * @type string
+	 */
+	const FLAG_PREVIOUS_ORDER_PAID = 'wcpay_paid_for_previous_order';
+
+	/**
+	 * Flag to indicate that a previous intention attached to the order was successful.
+	 *
+	 * @type string
+	 */
+	const FLAG_PREVIOUS_SUCCESSFUL_INTENT = 'wcpay_previous_successful_intent';
+
+	/**
+	 * OrderService instance.
+	 *
+	 * @var OrderService
+	 */
+	private $order_service;
+
+	/**
+	 * HooksProxy instance.
+	 *
+	 * @var HooksProxy
+	 */
+	private $hooks_proxy;
+
+	/**
+	 * LegacyProxy instance.
+	 *
+	 * @var LegacyProxy
+	 */
+	private $legacy_proxy;
+
+	/**
+	 * Session service instance.
+	 *
+	 * @var SessionService
+	 */
+	private $session_service;
+
+	/**
+	 * Logger instance.
+	 *
+	 * @var Logger
+	 */
+	private $logger;
+
+	/**
+	 * Initializes all dependencies.
+	 *
+	 * @param  OrderService   $order_service   Order service instance.
+	 * @param  SessionService $session_service Session service instance.
+	 * @param  Logger         $logger          Logger instance.
+	 * @param  HooksProxy     $hooks_proxy     Hooks proxy instance.
+	 * @param  LegacyProxy    $legacy_proxy    Legacy proxy instance.
+	 */
+	public function __construct( OrderService $order_service, SessionService $session_service, Logger $logger, HooksProxy $hooks_proxy, LegacyProxy $legacy_proxy ) {
+		$this->order_service   = $order_service;
+		$this->session_service = $session_service;
+		$this->logger          = $logger;
+		$this->hooks_proxy     = $hooks_proxy;
+		$this->legacy_proxy    = $legacy_proxy;
+	}
+
+	/**
+	 * Initializes this class's hooks.
+	 *
+	 * @return void
+	 */
+	public function init_hooks(): void {
+		// Priority 21 to run right after wc_clear_cart_after_payment.
+		$this->hooks_proxy->add_action( 'template_redirect', [ $this, 'clear_session_processing_order_after_landing_order_received_page' ], 21 );
+	}
+
+	/**
+	 * Checks if the currently attached payment intent was authorized for the current processing order.
+	 *
+	 * @param  int $order_id ID of the current processing order.
+	 *
+	 * @return WC_Payments_API_Payment_Intention|null The authorized attached payment intent, null otherwise.
+	 * @throws Order_Not_Found_Exception
+	 */
+	public function get_authorized_payment_intent_attached_to_order( int $order_id ): ?WC_Payments_API_Payment_Intention {
+		$intent_id = $this->order_service->get_intent_id( $order_id );
+		if ( is_null( $intent_id ) ) {
+			return null;
+		}
+
+		// We only care about payment payment_intent.
+		$is_payment_intent = 'pi_' === substr( $intent_id, 0, 3 );
+		if ( ! $is_payment_intent ) {
```

Comment:
Nitpick: Could we avoid the variable, and use the condition straight within `if()`?

Reply:
Did it c7de0d1adf9a7ac47dc05ffb44cb7d42d4b2aaa9